### PR TITLE
Replace the quadradic duplicate check logic with a linear version

### LINKED ISSUE #68: Quadratic duplicate message check
Right now, the "fallback" duplicate check (when blst is disabled) is has a `n^2` runtime. The `blst` version is using a red/black tree and has a `log(n)` runtime.

I'd consider:

1. At a minimum, making the fallback have a `log(n)` runtime (e.g., use `BTreeSet`).
2. Even better, use a `HashSet` for both. The rust `HashSet` is pretty damn fast and avoids hashing entirely for small sets, IIRC, so we shouldn't have any issues there.

### PR DESCRIPTION
**Pull request by Stebalien**
It's ~2% slower with 10 hashes and ~2% faster with 100 hashes (with very noisy benchmarks). But the important part is that the runtime should scale linearly with the number of inputs. E.g., this version is 2.5x faster for 1000 signatures.

fixes #68

### CODE REVIEW

**Comment by Stebalien** [comment]
Based on #67.

File: src/signature.rs [hunk 1]
```diff
@@ -139,12 +139,15 @@ pub fn verify(signature: &Signature, hashes: &[G2Projective], public_keys: &[Pub
...
-    for i in 0..(n_hashes - 1) {
-        for j in (i + 1)..n_hashes {
-            if hashes[i] == hashes[j] {
-                return false;
-            }
-        }
+    if hashes
+        .iter()
+        // This is the best way to get something we can actually hash.
+        .map(|h| G2Affine::from(h).to_compressed())
```
Comment (DrPeterVanNostrand) [inline src/signature.rs:145]: Maybe using `G2Projective::to_bytes` (which does exactly the same thing as what you added) would make it totally clear that you're comparing bytes, and that the point representation (affine/projective compressed/uncompressed) doesn't affect duplicate checking?
```rust
// Convert G2 point to a type that implements `Hash`.
.map(G2Projective::to_bytes)
```

**Review by DrPeterVanNostrand — CHANGES_REQUESTED** [review]

File: src/signature.rs (hunk 1 above)
Comment (Stebalien) [reply src/signature.rs:145]: Unfortunately, that returns a `G2Compressed` which doesn't let me get at the underlying `[u8; 96]`. I can call `as_ref` on that, but then I'd need to copy the bytes out into a new buffer and deal with all the potential error cases (and/or allocate).

I'm happy to do that if you think it'll work better, but it'll end up being:

```rust
<[u8; 96]>::try_from(h.to_bytes().as_ref()).unwrap()
```

Comment (Stebalien) [reply src/signature.rs:145]: I'll see if I can get upstream to accept an `Into<[u8; 96]>` patch.

Comment (Stebalien) [reply src/signature.rs:145]: https://github.com/zkcrypto/bls12_381/pull/119. But in the meantime, I'm happy to do the copy dance if you think it's easier to understand.

Comment (DrPeterVanNostrand) [reply src/signature.rs:145]: Ha oh right, `to_bytes` doesn't actually return the array of bytes... so I think how it's currently written is best.

**Review by DrPeterVanNostrand — APPROVED** [review]

**Pull request merged**